src/main/java/yeelp/distinctdamagedescriptions/util/lib/damagecalculation/DDDCombatTracker.java

```diff
@@ -126,14 +120,14 @@ public void handleAttackStage(LivingAttackEvent evt) {
 
 	public void handleHurtStage(LivingHurtEvent evt) {
 		this.updateContextAndDamage(evt.getSource(), evt.getAmount(), evt.getSource().getImmediateSource());
-		this.getIncomingDamage().filter(Predicates.not(Map::isEmpty)).ifPresent((m) -> {
-			if(m.size() < 1)return;
+		this.getIncomingDamage().filter(((Predicate<DamageMap>) Map::isEmpty).negate()).ifPresent((m) -> {
 			this.getCurrentlyUsedShieldDistribution().ifPresent((shield) -> {
 				shield.block(m);
 				this.results.hasEffectiveShield(m);
 				this.ctx.getShield().ifPresent((stack) -> stack.damageItem((int) (evt.getAmount() * (this.getRecentResults().getShieldRatio().getAsDouble())), this.getFighter()));
 			});
-			this.type = m.keySet().stream().collect(Collectors.toList()).get(rand.nextInt(m.size()));
+			this.type = new ArrayList<>(m.keySet()).get(
```

Comment:
We should add a size check here. I think something like

```
if(!m.isEmpty()) {
   this.type = m.keySet().stream().skip(rand.nextInt(m.size()).findFirst().get();
}
```

actually works well and solves the problem. If the damage is completely blocked and the map is empty, `this.type` is perfectly fine to be `null`, as this member is only used for death messages (and if the map is empty, the player won't die during this damage calculation). Plus, `this.type` gets wrapped in an Optional anyway. 

I'm not sure which is better, using `Set.stream()` with `Stream.skip(long)` or initializing a whole ArrayList for a single `get(int)` operation. Initializing the ArrayList iterates over the entire set. `Set.stream()` may not. Some research also suggests using an Iterator to skip the first n items could also work.

Reply:
Also, I should mention this is why blocked damage needs to be removed from the map. For example, if an attack inflicts piercing and bludgeoning, and you block the piercing fully but the bludgeoning damage kills the player, the death message shouldn't say that piercing damage killed the player, because it didn't. Hopefully that explains it well.

Reply:
Make sense, I will replace it with size check. The change to Stream is come from IDEA and I mistakenly add it to git, I will revert it. There are many Optional != null thing in the same file, those should be what really need to change, maybe initialize it to () -> null or similar